### PR TITLE
fix: element injector vs module injector

### PR DESCRIPTION
**Pull request by vicb**
TODO:
- [x] add refs to issues in commit message,
- [x] describe API change in commit message
- [x] link design doc
- [x] test for `ViewContainerRef.parentInjector`

### CODE REVIEW

File: modules/@angular/core/src/application_ref.ts [hunk 1]
```diff
@@ -502,7 +501,7 @@ export class ApplicationRef_ extends ApplicationRef {
       componentFactory = this._componentFactoryResolver.resolveComponentFactory(componentOrFactory);
     }
     this._rootComponentTypes.push(componentFactory.componentType);
-    const compRef = componentFactory.create(this._injector, [], componentFactory.selector);
+    const compRef = componentFactory.create(Injector.NULL, [], componentFactory.selector);
```
Comment (vicb) [inline modules/@angular/core/src/application_ref.ts]: explicitely pass the module ?

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 2]
```diff
@@ -81,34 +79,35 @@ export class NgComponentOutlet implements OnChanges, OnDestroy {
...
-      let injector = this.ngComponentOutletInjector || this._viewContainerRef.parentInjector;
+      const elInjector = this.ngComponentOutletInjector || this._viewContainerRef.parentInjector;
 
-      if ((changes as any).ngComponentOutletNgModuleFactory) {
+      if (changes['ngComponentOutletNgModuleFactory']) {
         if (this._moduleRef) this._moduleRef.destroy();
+
         if (this.ngComponentOutletNgModuleFactory) {
-          this._moduleRef = this.ngComponentOutletNgModuleFactory.create(injector);
+          const ngModuleRef = elInjector.get(NgModuleRef);
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: parentModule

File: modules/@angular/core/src/linker/component_factory_resolver.ts [hunk 3]
```diff
@@ -44,18 +44,32 @@ export abstract class ComponentFactoryResolver {
...
     }
   }
 
   resolveComponentFactory<T>(component: {new (...args: any[]): T}): ComponentFactory<T> {
-    let result = this._factories.get(component);
-    if (!result) {
-      result = this._parent.resolveComponentFactory(component);
-    }
-    return result;
+    let factory = this._factories.get(component) || this._parent.resolveComponentFactory(component);
```
Comment (vicb) [inline modules/@angular/core/src/linker/component_factory_resolver.ts]: add tests

File: modules/@angular/core/src/application_ref.ts (hunk 1 above)
Comment (vicb) [inline modules/@angular/core/src/application_ref.ts]: pass the module explicitely

Comment (vicb) [inline modules/@angular/core/src/application_ref.ts]: test:
- create a child module and bootstrap a cmp from it, should see child providers.

File: modules/@angular/core/src/linker/component_factory_resolver.ts [hunk 4]
```diff
@@ -44,18 +44,32 @@ export abstract class ComponentFactoryResolver {
...
+
+    return factory ? new ComponentFactoryBoundToModule(factory, this._ngModule) : null;
+  }
+}
+
+export class ComponentFactoryBoundToModule<C> extends ComponentFactory<C> {
+  constructor(private factory: ComponentFactory<C>, private ngModule: NgModuleRef<any>) { super(); }
```
Comment (vicb) [inline modules/@angular/core/src/linker/component_factory_resolver.ts]: ngModule should be public on `ComponentFactory`

File: modules/@angular/core/src/view/refs.ts [hunk 5]
```diff
@@ -52,12 +53,13 @@ class ComponentFactory_ extends ComponentFactory<any> {
    * Creates a new component.
    */
   create(
-      injector: Injector, projectableNodes: any[][] = null,
-      rootSelectorOrNode: string|any = null): ComponentRef<any> {
+      injector: Injector, projectableNodes?: any[][], rootSelectorOrNode?: string|any,
+      ngModule?: NgModuleRef<any>): ComponentRef<any> {
```
Comment (vicb) [inline modules/@angular/core/src/view/refs.ts]: throw if no ngModule

File: modules/@angular/core/src/view/refs.ts [hunk 6]
```diff
@@ -52,12 +53,13 @@ class ComponentFactory_ extends ComponentFactory<any> {
...
     const viewDef = resolveViewDefinition(this.viewDefFactory);
     const componentNodeIndex = viewDef.nodes[0].element.componentProvider.index;
+    const module = ngModule || injector.get(NgModuleRef);
```
Comment (vicb) [inline modules/@angular/core/src/view/refs.ts]: revert

File: modules/@angular/core/src/view/refs.ts [hunk 7]
```diff
@@ -105,7 +107,8 @@ class ViewContainerRef_ implements ViewContainerRef {
       elDef = viewParentEl(view);
       view = view.parent;
     }
-    return view ? new Injector_(view, elDef) : this._view.root.injector;
+
+    return view ? new Injector_(view, elDef) : Injector.NULL;
```
Comment (vicb) [inline modules/@angular/core/src/view/refs.ts]: revert
add test: inject `ViewContainerRef` on a cmp TestBed.create(Comp).get()

File: modules/@angular/core/test/linker/ng_module_integration_spec.ts [hunk 8]
```diff
@@ -130,9 +130,13 @@ function declareTests({useJit}: {useJit: boolean}) {
     }
 
     function createComp<T>(compType: Type<T>, moduleType: Type<any>): ComponentFixture<T> {
-      const ngModule = createModule(moduleType);
+      const ngModule = createModule(moduleType, TestBed.get(Injector));
```
Comment (vicb) [inline modules/@angular/core/test/linker/ng_module_integration_spec.ts]: TestBed.get() -> injector

File: modules/@angular/router/src/apply_redirects.ts [hunk 9]
```diff
@@ -280,7 +282,9 @@ class ApplyRedirects {
 
   private getChildConfig(injector: Injector, route: Route): Observable<LoadedRouterConfig> {
```
Comment (vicb) [inline modules/@angular/router/src/apply_redirects.ts]: check if we can pass the module instead

File: modules/@angular/router/src/router.ts [hunk 10]
```diff
@@ -991,7 +991,7 @@ export class PreActivation {
 
   private getToken(token: any, snapshot: ActivatedRouteSnapshot): any {
     const config = closestLoadedConfig(snapshot);
-    const injector = config ? config.injector : this.injector;
+    const injector = config ? config.module.injector : this.injector;
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: injector -> ModuleRef only

File: modules/@angular/router/src/router.ts [hunk 11]
```diff
@@ -1102,22 +1102,19 @@ class ActivateRoutes {
...
 
     if (config) {
-      injector = config.injectorFactory(outlet.locationInjector);
-      resolver = config.factoryResolver;
-      resolved.push({provide: ComponentFactoryResolver, useValue: resolver});
+      resolver = config.module.componentFactoryResolver;
     } else {
-      injector = outlet.locationInjector;
       resolver = outlet.locationFactoryResolver;
     }
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: add simpler method & deprecate

` outlet.activate(future, resolver, outletMap);`

File: modules/@angular/router/src/router_preloader.ts [hunk 12]
```diff
@@ -91,37 +91,40 @@ export class RouterPreloader {
     this.subscription = concatMap.call(navigations, () => this.preload()).subscribe(() => {});
   }
 
-  preload(): Observable<any> { return this.processRoutes(this.injector, this.router.config); }
+  preload(): Observable<any> {
+    const ngModule = this.injector.get(NgModuleRef);
```
Comment (vicb) [inline modules/@angular/router/src/router_preloader.ts]: inject NgModuleRef

File: modules/@angular/router/test/integration.spec.ts [hunk 13]
```diff
@@ -2422,6 +2422,175 @@ describe('Integration', () => {
              expect(fixture.nativeElement).toHaveText('lazy-loaded-parent [lazy-loaded-child]');
            })));
 
+    it('should have 2 injector trees: module and element',
+       fakeAsync(inject(
+           [Router, Location, NgModuleFactoryLoader],
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: add module can no longer shadow el

File: modules/@angular/common/src/directives/ng_component_outlet.ts (hunk 2 above)
Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: done

File: modules/@angular/core/src/linker/component_factory_resolver.ts (hunk 3 above)
Comment (vicb) [inline modules/@angular/core/src/linker/component_factory_resolver.ts]: done

File: modules/@angular/core/src/application_ref.ts (hunk 1 above)
Comment (vicb) [inline modules/@angular/core/src/application_ref.ts]: done

File: modules/@angular/core/src/linker/component_factory_resolver.ts (hunk 4 above)
Comment (vicb) [inline modules/@angular/core/src/linker/component_factory_resolver.ts]: not for now

File: modules/@angular/core/src/view/refs.ts (hunk 6 above)
Comment (vicb) [inline modules/@angular/core/src/view/refs.ts]: done

File: modules/@angular/core/test/linker/ng_module_integration_spec.ts (hunk 8 above)
Comment (vicb) [inline modules/@angular/core/test/linker/ng_module_integration_spec.ts]: done

File: modules/@angular/router/src/router.ts (hunk 11 above)
Comment (vicb) [inline modules/@angular/router/src/router.ts]: done

File: modules/@angular/router/src/router_preloader.ts (hunk 12 above)
Comment (vicb) [inline modules/@angular/router/src/router_preloader.ts]: no, API break

File: modules/@angular/router/test/integration.spec.ts (hunk 13 above)
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: done

File: modules/@angular/router/src/router.ts (hunk 10 above)
Comment (vicb) [inline modules/@angular/router/src/router.ts]: no -> API break

File: modules/@angular/core/src/view/refs.ts [hunk 14]
```diff
@@ -55,11 +55,13 @@ class ComponentFactory_ extends ComponentFactory<any> {
   create(
       injector: Injector, projectableNodes?: any[][], rootSelectorOrNode?: string|any,
       ngModule?: NgModuleRef<any>): ComponentRef<any> {
+    if (!ngModule) {
+      throw new Error('ngModule should be provided');
```
Comment (tbosch) [inline modules/@angular/core/src/view/refs.ts]: Maybe `Illegal sate: NgModule has to be provided`?

Comment (tbosch) [inline modules/@angular/core/src/view/refs.ts]: (as its an error that users never should run into)...

**Review by tbosch — APPROVED** [review]
Reviewed in person...

**Comment by thaoula** [comment]
hi vicb,

Thanks for your hard work. 

Do you have any idea when this will be merged into V4?

Regards,
Tarek

**Review by IgorMinar — APPROVED** [review]
lgtm for api changes

**Comment by vicb** [comment]
@thaoula should be part of the next RC release

**Comment by thaoula** [comment]
Hi Vicb, 

Awesome!!!. Do you know when that will be? Been waiting for a fix to this issue since November.

Regards,
Tarek

File: packages/core/src/view/refs.ts [hunk 15]
```diff
@@ -52,12 +53,15 @@ class ComponentFactory_ extends ComponentFactory<any> {
    * Creates a new component.
    */
   create(
-      injector: Injector, projectableNodes: any[][] = null,
-      rootSelectorOrNode: string|any = null): ComponentRef<any> {
+      injector: Injector, projectableNodes?: any[][], rootSelectorOrNode?: string|any,
+      ngModule?: NgModuleRef<any>): ComponentRef<any> {
+    if (!ngModule) {
```
Comment (apoterenko) [inline packages/core/src/view/refs.ts]: hi!

ngModule is defined as an optional variable, but here we see throw instruction.

Why?

Comment (chuckjaz) [inline packages/core/src/view/refs.ts]: This class implements the `ComponentFactory<any>` which requires this parameter be optional (https://github.com/angular/angular/blob/4.0.0-rc.4/packages/core/src/linker/component_factory.ts#L76-L78). However, this is an internal implementation that should never be called without a module so we check for internal consistency.

**Pull request merged**